### PR TITLE
Move global css into AppProvider

### PR DESCRIPTION
**Pull request by BPScott**
## WHY are these changes introduced?

Consuming apps that use the esnext build have to import a global.scss file. This is distinct from all our other styles that get imported as side effects of using particular components. Prior to sewing-kit v0.113.0 this was handled automatically by sewing-kit, however as part of my push to remove polaris-specific config from sewing-kit (and thus remove the coupling of "you must use sewing-kit vX with polaris-react vY") v0.113.0 onwards will require [developers to explictly import this file at the same place where they import the AppProvider](https://github.com/Shopify/sewing-kit/pull/1599), per SK's [sass documentation](https://github.com/Shopify/sewing-kit/blob/master/docs/plugins/sass.md#using-polaris-and-other-component-libraries).

What if there was a way to remove that manual step?

### WHAT is this pull request doing?

This PR moves our global scss to be part of the AppProvider component.

Instead of importing a global.scss file, consuming applications get the global styles as a side-effect of importing AppProvider. 

This has no effect for consumers using our regular builds as the compiled styles.css remains identical.

For those using our esnext build (i.e. people using sewing-kit), as AppProvider is a required wrapper component if you want to use Polaris you are already including his on your pages. This removing the need to specify an explict `import '@shopify/polaris/esnext/styles/global.scss';` in the consuming app.

An empty global.scss file remains present in order to not break consuming applications that are on older versions of sewing-kit which expect the global.scss to exist. This file can be removed in v6 at which point we say "you use polaris v6 you must use at least sewing-kit v0.113.0".

### How to 🎩

- build consumer into the web project and ensure global styles for font-sizing are still present.
- do a build and confirm styles/components.scss contains the html and body stylings at the top of the file.

### CODE REVIEW

**Comment by BPScott** [comment]
@tmlayton and @danrosenthal, @GoodForOneFare, @lemonmade can you think of any reason why this may cause problems with esnext builds in sewing-kit?

**Comment by BPScott** [comment]
(also yo @beefchimi / @tsov you might be interested in this)

**Comment by GoodForOneFare** [comment]
I would figure out how apps with odd Polaris configs would be affected by this change:
- [pay](https://github.com/Shopify/pay/blob/master/config/sewing-kit.config.ts#L35)
- [u2](https://github.com/Shopify/u2/blob/master/config/sewing-kit.config.ts#L34)
- [exchange](https://github.com/Shopify/exchange/blob/master/config/sewing-kit.config.js#L54)
- [shopify-app-store](https://github.com/Shopify/shopify-app-store/blob/master/config/sewing-kit.config.ts#L8)
- [partners](https://github.com/Shopify/partners/blob/master/config/sewing-kit.config.ts#L61)
- [help](https://github.com/Shopify/help/blob/master/config/sewing-kit.config.js#L92)

**Comment by lemonmade** [comment]
My main concern is that it will be shaken out of production builds because it is only imported for side effects. Right now, I *think* it is only preserved because it is explicitly listed as part of the entry array.

**Comment by BPScott** [comment]
> My main concern is that it will be shaken out of production builds because it is only imported for side effects

Does that mean that the instruction at https://github.com/Shopify/sewing-kit/blob/master/docs/plugins/sass.md#using-polaris-and-other-component-libraries for how to manually import the current global file won't work in production? If so what do you think would be the best way to fix it? Telling people to do the following feels kinda janky:

```ts
@ts-ignore
import * as globalStyles from '@shopify/polaris/esnext/styles/global.scss';
```

**Comment by lemonmade** [comment]
I think they would need to add the Polaris Sass file to the list of sideEffects in package.json, but that is just a guess. IMO that’s not really an acceptable DX so we would have to find a way around it if it’s the case.

**Comment by BPScott** [comment]
> My main concern is that it will be shaken out of production builds because it is only imported for side effects. 

We do an import for side-effects only in ScrollLock and that CSS seems to be present (I've not verified this but I assume it works because nobody has pointed out scrollable BGs behind modals)
https://github.com/Shopify/polaris-react/blob/master/src/components/ScrollLock/ScrollLock.ts#L3

**Comment by BPScott** [comment]
Investigation time: 

Using Polaris without importing our global scss isn't supported and it's only through dumb luck that apps that do so continue to render correctly.

- [pay](https://github.com/Shopify/pay/blob/master/config/sewing-kit.config.ts#L35)
  - Pay doesn't import polaris's global styles, and instead renders their own https://github.com/Shopify/pay/blob/master/app/ui/styles/global.scss. They might need to make some modifications to ensure overriding where possible but this looks managable as they're mostly similar to ours. 
- [u2](https://github.com/Shopify/u2/blob/master/config/sewing-kit.config.ts#L34)
  - u2 imports the polaris global scss in varying entry points within the app (e.g. https://github.com/Shopify/u2/blob/fc3b991aee7fbdcaab5c82b9bde7614883ef70cb/app/ui/rails_frame/index.tsx) in the same locations that they import AppProvider. It looks like this will work for them.
- [exchange](https://github.com/Shopify/exchange/blob/master/config/sewing-kit.config.js#L54)
  - exchange imports the polaris global scss into a single entrypoint and within that entrypoint they render our AppProvider. This will work for them and they can remove logic from their sk config file.
- [shopify-app-store](https://github.com/Shopify/shopify-app-store/blob/master/config/sewing-kit.config.ts#L8)
    - app-store imports the polaris global scss into multiple entry points and within those entrypoint they render our AppProvider. They'll have to move their AppProvider and global.scss mentions closer together (as per SK 0.112.0 suggests) but this should work for them.
- [partners](https://github.com/Shopify/partners/blob/master/config/sewing-kit.config.ts#L61)
  - partners do a bunch of webpack magic to entirely opt out of tree shakable scss, and instead import plain old `@shopify/polaris/styles.scss`. This will work for them the full build will still contain this file 
- [help](https://github.com/Shopify/help/blob/master/config/sewing-kit.config.js#L92)
  - I can't tell where help opts into our global styles at all :/


With all that said, this feels pretty scary for a minor release, I'm going to rebase this atop the v5 branch and heartily suggest people are using at least sewing-kit 0.113.0 before updating to polaris v5. This will work with older versions but it's nice to give a "this will be least complicated" cutoff.

**Comment by BPScott** [comment]
Rebased atop version-5.0.0. This is now ready for review.

File: src/styles/shared/_skeleton.scss [hunk 1]
```diff
@@ -18,7 +18,7 @@ $thumbnail-sizes: (
 
 @mixin skeleton-shimmer {
   :global {
-    // This is a global animation, defined in /src/styles/global.scss
+    // This is a global animation, defined in /src/components/AppProvider/AppProvider.scss
```
Comment (GoodForOneFare) [inline src/styles/shared/_skeleton.scss]: If I had one wish for Polaris 5, it'd be 🔥 the shimmer 😈

File: src/components/index.ts [hunk 2]
```diff
@@ -1,11 +1,13 @@
+// AppProvider contains CSS that affects element level CSS (e.g. `html`, `button`)
+// It should be first to ensure its CSS is first in compiled output
+export {AppProvider, AppProviderProps} from './AppProvider';
```
[the remark this answers is not in the document]
Comment (BPScott) [reply src/components/index.ts]: Yeah I believe so as they'd appear below the styles that get pulled out from AppProvider.

Worst case they can resort to `!important` or do obtuse things with specificity if they need to trump polaris e.g.`html:root` will trump `html` but isn't the nuclear tier of`!important`.

**Comment by BPScott** [comment]
I'm going to leave that global.scss file present for the moment, so apps using SK <0.113.0 can use this without needing to modify their webpack config. We can kill it in v6

**Pull request merged**